src/spetlr/power_bi/PowerBi.py

```diff
@@ -0,0 +1,436 @@
+                    + " (local time)"
+                )
+                if (
+                    self.max_minutes_after_last_refresh > 0
+                    and self.last_refresh_utc < self.min_refresh_time_utc
+                ):
+                    self._raise_error(
+                        "Last refresh finished more than "
+                        f"{self.max_minutes_after_last_refresh} "
+                        f"minutes ago at {last_refresh_str} !"
+                    )
+                else:
+                    print(f"Refresh completed successfully at {last_refresh_str}.")
+                    return True
+        elif self.last_status == "Unknown":
+            self._raise_error("Refresh is still in progress!")
+        elif self.last_status == "Disabled":
+            self._raise_error("Refresh is disabled!")
+        elif self.last_status == "Failed":
+            self._raise_error(f"Last refresh failed! {self.last_exception}")
+        else:
+            self._raise_error(
+                f"Unknown refresh status: {self.last_status}! {self.last_exception}"
+            )
+        return False
+
+    def _trigger_new_refresh(self) -> bool:
+        """
+        Starts a refresh of the PowerBI dataset.
+
+        :return: True if succeeded or False if failed (when ignore_errors==True)
+        :rtype: bool
+        :raises Exception: if failed and ignore_errors==False
+        """
+
+        if self.last_status is None or self.last_status in ["Completed", "Failed"]:
+            if self.last_status == "Failed":
+                print(f"Warning: Last refresh failed! {self.last_exception}")
+                print()
+
+            api_url = (
+                f"{self.powerbi_url}groups/{self.workspace_id}"
+                f"/datasets/{self.dataset_id}/refreshes"
+            )
+            api_call = requests.post(url=api_url, headers=self.api_header)
+            if api_call.status_code == 202:
+                print("A new refresh has been successfully triggered.")
+                return True
+            else:
+                self._raise_api_error("Failed to trigger a refresh!", api_call)
+        elif self.last_status == "Unknown":
+            print("Refresh is already in progress!")
+            return True
+        elif self.last_status == "Disabled":
+            self._raise_error("Refresh is disabled!")
+        else:
+            self._raise_error(
+                f"Unknown refresh status: {self.last_status}! {self.last_exception}"
+            )
+        return False
+
+    def _get_seconds_to_wait(self, elapsed: int) -> int:
+        """
+        Returns the number of seconds to wait before rechecking
+            if the refresh has completed.
+
+        :return: number of seconds to wait
+        :rtype: int
+        """
+
+        if self.last_duration > 0:
+            elapsed = abs(self.last_duration - elapsed)
+
+        return 15 if elapsed < 40 else 60 if elapsed < (3 * 60) else (5 * 60)
```

Comment:
Where comes this logic from. I don't understand.

Reply:
Microsoft set a limit of how many API requests in an hour can be called. So this logic is a fine compromise between how often we check if the refresh finished (small delay) and how late we finish checking (large delay).

Reply:
If the time elapsed is less than 40 seconds, wait 15 seconds.
If the elapsed time is between 40 seconds and 3 minutes, wait 60 seconds
Else wait 5 minutes?